web3.js/src/connection.ts

```diff
@@ -3154,6 +3176,143 @@ export class Connection {
+    signature: string;
+    strategy: NonceBasedTransactionConfirmationStrategy;
+  }) {
+    let done: boolean = false;
+    const expiryPromise = new Promise<{
+      __type: TransactionStatus.NONCE_INVALID;
+    }>(resolve => {
+      const INVALID_NONCE_ERROR_SENTINEL = {} as const;
+      let currentNonceValue = nonceValue;
+      const getCurrentNonceValue = async () => {
+        try {
+          const nonceAccount = await this.getNonce(nonceAccountPubkey, {
+            commitment,
+            minContextSlot,
+          });
+          if (nonceAccount == null) {
+            throw INVALID_NONCE_ERROR_SENTINEL;
+          }
+          return nonceAccount.nonce;
+        } catch (e) {
+          if (e === INVALID_NONCE_ERROR_SENTINEL) {
+            throw e;
+          }
+          // If for whatever reason we can't reach/read the nonce
+          // account, just keep using the last-known value.
+          return currentNonceValue;
+        }
+      };
+      (async () => {
+        try {
+          currentNonceValue = await getCurrentNonceValue();
+          if (done) return;
+          while (
+            true // eslint-disable-line no-constant-condition
+          ) {
+            if (nonceValue !== currentNonceValue) {
+              resolve({__type: TransactionStatus.NONCE_INVALID});
+              return;
+            }
+            await sleep(2000);
+            if (done) return;
+            currentNonceValue = await getCurrentNonceValue();
+            if (done) return;
+          }
+        } catch (e) {
+          if (e === INVALID_NONCE_ERROR_SENTINEL) {
+            resolve({__type: TransactionStatus.NONCE_INVALID});
+            return;
+          } else {
+            throw e;
+          }
+        }
+      })();
+    });
+    const [confirmationPromise, cancelConfirmationDetector] =
+      this.getTransactionConfirmationPromise({commitment, signature});
+    let result: RpcResponseAndContext<SignatureResult>;
+    try {
+      const outcome = await Promise.race([confirmationPromise, expiryPromise]);
+      if (outcome.__type === TransactionStatus.PROCESSED) {
+        result = outcome.response;
+      } else {
+        // Double check that the transaction is indeed unconfirmed.
+        let signatureStatus:
+          | RpcResponseAndContext<SignatureStatus | null>
+          | null
+          | undefined;
+        while (
+          true // eslint-disable-line no-constant-condition
+        ) {
+          const status = await this.getSignatureStatus(signature);
+          if (status == null) {
+            break;
+          }
+          if (status.context.slot < minContextSlot) {
+            await sleep(400);
+            continue;
+          }
+          signatureStatus = status;
+          break;
+        }
+        if (signatureStatus?.value) {
+          const commitmentForStatus = commitment || 'finalized';
+          const {confirmationStatus} = signatureStatus.value;
+          switch (commitmentForStatus) {
```

Comment:
What is the purpose of this commitment comparison logic? Shouldn't `getSignatureStatus` be called with the same commitment that is used in expiration?

Reply:
This code says:

1. So you received a response that says the signature has been processed with finality _x_.
2. Does finality _x_ meet _at least_ the threshold of commitment used for `confirmTransaction`?

Reply:
So for instance, if you used `finalized` with `confirmTransaction` but discovered that the signature status has finality `processed` – no deal. That's a failure.

Reply:
Ah, I see. I didn't realize that `getSignatureStatus` can't be called with a `commitment` at all. Most APIs allow you to make a query with a `commitment` and `minContextSlot` that tells the caller the state of the world for a block of a particular commitment level whose slot is greater than minContextSlot which is what we want here. I wish I could suggest using `getTransaction` API instead but it doesn't support the `minContextSlot` param and it also doesn't return the context slot in which it was evaluated either.

It's difficult to rely on `confirmationStatus` to make any determinations about what a transaction's comittment will eventually be because each node has its own view on what the finalization status of a particular block is and could be lagging behind other nodes.

> So for instance, if you used finalized with confirmTransaction but discovered that the signature status has finality processed – no deal. That's a failure.

So in this example, the signature status is currently only "processed" but could eventually be finalized if the block at `context.slot` gets finalized or if the node finalizes the transaction in another fork.

I think the best thing that we can do right now is a batch request to `getTransaction` and `getSlot` and use `commitment` for each of them. If the returned slot from `getSlot` is high enough, we can be confident that node definitely didn't process the transaction if `getTransaction` returns null.

Reply:
cc @CriesofCarrots, we want to be able to use the RPC API here to know for sure that a transaction wasn't included before a particular slot with a particular commitment. Any ideas besides my suggestion to batch `getTransaction` and `getSlot` in a batch request? I really wish `getTransaction` returned a context wrapped response... the lack of that is why `minContextSlot` wasn't added to this API.

Reply:
Another thing we could do is use the `getSignaturesForAddress` API called with the nonce account key. That API supports both `commitment` and `minContextSlot`

Reply:
I created an issue to track this problem here: https://github.com/solana-labs/solana/issues/27098

Reply:
I still _think_ this works, however sub-optimal it might be.

1. You reach this code if the nonce was found to have advanced.
2. In the extremely unlikely event that the signature _subscription_ notification arrived _after_ the nonce was found to have advanced, this code aims to double-check whether the signature, in fact, was confirmed, or not.
3. The double-check makes sure to compare the signature status at the same slot as the nonce was found to advance.